Fts5 support

Corion opened the pull request:
This is not yet ready for merging, as it fails the new test on OSX!

I'm opening the pull request to get insights on how to better structure this and put the rudimentary implementation out there. I have not used the code beyond its simple self-test.

This needs at least the following things done:

- [ ] Add documentation for using custom FTS5 tokenizers
- [x] Fix test error on OSX / clang - this is most likey an error in the added XS code that goes unnoticed on gcc
- [ ] Find a better API for `DBD::SQLite::db::fts5_xToken()` instead of calling it fully qualifed in `DBD::SQLite::db::`

Commit:
WIP: Stuff needed for FTS5

Commit:
Remove trailing whitespace

Commit:
Fix all type errors

Commit:
The test suite now passes without segfaults or anything

Commit:
Note an optimization, for later

Commit:
Silence a compilation warning on Mac/clang

Gcc doesn't warn on this (at my settings) but Github automation found
this one

Commit:
Eliminate unused (but returned) variable

Commit:
Eliminate call to strlen() in favour of SvPV()

damil commented:
Hi,
I'm happy to find out that you started working on FTS5 support for Perl. Do you intend to pursue this work ? If you need help I have a bit of time to volunteer for some tasks. Cheers, Laurent

Corion commented:
Yes, I really plan to continue work on this, but currently I'm looking for a good corpus to test with and some quiet time to actually work on the API. I think there are no problems on the XS side of the world anymore, but if you're interested, my current options for the Perl search API are:

1. Use an absolute name
```
sub locale_tokenizer { # see also: Search::Tokenizer
  return sub {
    my( $ctx, $string, $tokenizer_context_flags ) = @_;
    ...
    my $flags = FTS5_TOKEN_COLOCATED;
    DBD::SQLite::db::fts5_xToken($ctx,$flags,$term,$start,$end);
    ...
}
```

2. Pass in the tokenizer function to the callback

```
sub locale_tokenizer { # see also: Search::Tokenizer
  return sub {
    my( $ctx, $string, $tokenizer_context_flags ) = @_;
    ...
    my $flags = FTS5_TOKEN_COLOCATED;
    $ctx->token( $flags,$term,$start,$end );
    ...
}
```

I'm not happy with either approach, but I haven't written any code actually using it beyond `t/67_fts5.t`. I mildly favour the second approach, as it allows for more flexibility, but as you're the author of Search::Tokenizer, maybe you have more insight on that.

damil commented:
Hi, thanks for the answer.

Regarding the tokenizer API, I doubt that it could work by calling the FTS5 core from the perl sub; in FTS3 it's the other way around, and it seems to me that FTS5 has the same approach (although I havent't studied it yet in detail). This inverse calling pattern is necessary to be able to support several  tokenizers simultaneously. 
It works more or less like Russian puppets :

- FTS calls a Perl named sub to get a named tokenizer. This returns an anonymous sub.
-FTS calls  that sub on a given string, it gets back another anonymous sub
- the 2nd anonymous sub is an iterator; at each call from FTS it returns a new token, until reaching the end of the document where it returns NULL.

I had a quick look at fts5.h; the (*xToken) definition is quite close to (*xNext) in fts3_tokenizer.h, but not identical -- this means that the perl module Search::Tokenizer, which was designed for FTS4, cannot work as is for FTS5. It will need some adaptations, probably minor. Tell me what you need whenever you are ready.

Cheers, Laurent

Commit:
Rename "perl_tokenizer" to "perl_fts3_tokenizer"

This is because we will soon have two styles of tokenizers, one for
the FTS3 / FTS4 API and one for the FTS5 API

# Conflicts:
#	dbdimp_tokenizer.inc

Commit:
Properly export the FTS5_ constants on demand

Corion commented:
I've now used the FTS5 support in a small toy project ( [https://github.com/Corion/Filesys-DB] ) and while I still think the API is not great, it works as a first stab.

I think getting FTS5 enabled and supported in 1.72 would be great - what should I do to make this happen?

charsbar commented:
Thanks for the heads-up. I am not sure the current implementation is ready to be merged for a few reasons.

* The perl_fts5_tokenizer_Tokenize does not follow the pattern other bundled tokenizers do (cf. https://github.com/sqlite/sqlite/blob/master/ext/fts5/fts5_tokenize.c#L136-L167). xTokenize should return `a copy of xToken() return value` when something wrong happens (cf. the last paragraph of xTokenize section of https://sqlite.org/fts5.html#custom_tokenizers). As Laurent said previously, it is probably better for a perl tokenizer to return an iterator that returns values to pass to the xToken called in the main loop of perl_fts5_tokenizer_Tokenize. xToken callbacks are not something for custom tokenizer authors to handle by themselves.
* This PR is rather old and does not follow the recent changes about Unicode string (cf. https://github.com/DBD-SQLite/DBD-SQLite/pull/80). If we add some of the extra tests from t/43_fts.t (https://github.com/DBD-SQLite/DBD-SQLite/blob/master/t/43_fts3.t#L104-L137, excluding the ones for offsets() that fts5 does not have), the tests fail.

It would also be nice if dbdimp_tokenizer.inc is split into separate files (such as dbdimp_fts3_tokenizer.inc and dbdimp_fts5_tokenizer.inc) to make it easier to compare the counterparts.

Corion commented:
Thanks for the feedback, I'll look at things and implement them!

Commit:
Split up dbdimp_tokenizer.inc into FTS3 and FTS5

Commit:
Convert test file to UTF8 source code

Commit:
Also distribute fts5.h

Commit:
Strip whitespace

Commit:
Fix cpp comments test against "file:///"

Commit:
Move to v2 version of the tokenizer API

Maybe this helps us eliminate the global "locale" variable
we currently use to pass the document language to the tokenizer
implementations
